src/Components/Web.JS/src/Services/WebRootComponentManager.ts

```diff
@@ -215,10 +215,14 @@ export class WebRootComponentManager implements DescriptorHandler, RootComponent
     }, this._circuitInactivityTimeoutMs) as unknown as number;
   }
 
+  private rendererHasComponents(rendererId: WebRendererId): boolean {
+    const renderer = getRendererer(rendererId);
+    return renderer !== undefined && renderer.getRootComponentCount() > 0;
```

Comment:
when does getRootComponentCount gets populated? Can we be in a situation where we are starting both runtimes and end up having a similar situation? (Some components get assigned server, others webassembly, based on how quickly the runtimes start.

Also, what happens on situations where we start the server runtime, then navigate and kill it, and render new components? Do we switch to wasm at that point?

Reply:
> Can we be in a situation where we are starting both runtimes and end up having a similar situation?

Good thing to check! It looks like the server sends `'JS.AttachComponent'` with a descriptor, which triggers `attachRootComponentToLogicalElement` and thence `this.rootComponentIds.add(componentId)`. So the server renderer will only have components after the circuit is fully established, and not while we're still starting it.

So it does seem possible to have this sequence:

1. JS starts up and sees some auto component in the page
2. Without having the wasm files loaded, it resolves auto as `server`, so it starts establishing a circuit, sending the descriptor for the auto component
3. Before that process completes, the wasm files finish downloading and then the user does an enhanced nav to a new page that contains another auto component. That component would now resolve as `wasm`
4. Then the circuit from step 2 fully starts, and we attach its auto component

That would be much more of an edge case than before (where any enhanced nav after starting the circuit could resolve auto as wasm), but seems still possible.

There's already a `hasStartedServer` function that is used to make sure we don't start a 2nd circuit while the 1st one is still in process of starting, so maybe if `hasStartedServer` is true but `circuit.isDisposedOrDisposing` is false, we should resolve auto as server, regardless of whether the server renderer has any components, i.e., the priority order would be:

1. If wasm components already exist, resolve as wasm
2. If hasStartedServer && !circuit.isDisposedOrDisposing, resolve as server
3. If wasm resources are cached, resolve as wasm
4. Otherwise, resolve as server

Is this correct, @MackinnonBuck? You know this area better and maybe counting `server.rootComponentIds` is already equivalent.

Reply:
Good catch, and analysis seems correct to me. I think the right thing to do here would be to go with your suggestion and resolve Auto to use Blazor Server if the circuit is active, even if there aren't Server components on the page.

Reply:
I actually realized that this might lead to less desirable behavior if a page gets loaded with Auto, WebAssembly, and Server components all present. The logic above would resolve the Auto component to use Server because the circuit starts before the first WebAssembly component gets activated (considering the time it takes for the WebAssembly runtime to start). This actually broke an E2E test.

I'm proposing we use the following logic:
1. If there are any WebAssembly components, _or SSR components on the page that would get activated using WebAssembly_, resolve to WebAssembly
2. Same logic as above but for Server
3. If WebAssembly resources are cached, resolve as WebAssembly 
4. Otherwise, resolve as Server

The "or SSR components on the page that would get activated using WebAssembly" means:
* An SSR component exists in the document with the "server" or "webassembly" descriptor type, or...
* A component is in the middle of being activated for interactivity

^ This reuses part of the same logic we already use for determining whether the circuit can be closed.

How does this sound?